library/core/src/array/iter.rs

```diff
@@ -150,6 +150,27 @@ impl<T, const N: usize> Iterator for IntoIter<T, N> {
     fn last(mut self) -> Option<Self::Item> {
         self.next_back()
     }
+
+    fn advance_by(&mut self, n: usize) -> Result<(), usize> {
+        let len = self.len();
+
+        // The number of elements to drop.  Always in-bounds by construction.
+        let delta = cmp::min(n, len);
+
+        let range_to_drop = self.alive.start..(self.alive.start + delta);
+
+        // Moving the start marks them as conceptually "dropped", so if anything
+        // goes bad then our drop impl won't double-free them.
+        self.alive.start += delta;
+
+        // SAFETY: These elements are currently initialized, so it's fine to drop them.
+        unsafe {
+            let slice = self.data.get_unchecked_mut(range_to_drop);
+            ptr::drop_in_place(MaybeUninit::slice_assume_init_mut(slice));
+        }
```

Comment:
Tangent: These these dances seem a bit unergonomic and repeated in several places.

`ptr::drop_in_place` can take a `*mut [T]`, so for unsafe code it would be great if we had something like `[T].as_slice_ptr().get_unchecked(range).cast()` then we wouldn't have to refer to use those unwieldy associated methods on `MaybeUninit` since we're operating on pointers anyway. Generally there's a lack of helpers to work with `*mut/const [T]`.

Reply:
Hmm, there's `assume_init_drop` but no slice version of that, so I guess there could be `assume_init_drop_range` like there's `slice_assume_init_mut` and such -- one could always pass `..` to drop everything easily.

Don't think I'll do that in this PR, though.